src/_pytest/mark/structures.py

```diff
@@ -283,28 +284,38 @@ def test_function():
     on the ``test_function`` object. """
 
     _config = None
+    _markers = set()
 
     def __getattr__(self, name):
         if name[0] == "_":
             raise AttributeError("Marker name must NOT start with underscore")
+
         if self._config is not None:
-            self._check(name)
-        return MarkDecorator(Mark(name, (), {}))
+            # We store a set of markers as a performance optimisation - if a mark
+            # name is in the set we definitely know it, but a mark may be known and
+            # not in the set.  We therefore start by updating the set!
```

Comment:
When can it happen that it is known, but not in the set?  (except for the first read of the config of course)

(I've notied that the config is re-processed again for every unknown mark, whereas previously it was "read" just once)

Reply:
I think this covers the case where marks can be registered later, for example by `pytest_configure` hooks in `conftests.py` not at the root.

Reply:
Yep, `pytest_configure` can run later than I had expected.

The re-processing isn't free, but the number of unknown marks should be relatively small and it's necessary for correctness.